src/data/upload.ts

```diff
@@ -48,33 +55,51 @@ export function end(): void {
     clearTimeout(timeout);
     upload(true);
     events = [];
+    transit = {};
+    track = null;
 }
 
 function upload(last: boolean = false): void {
+    metric.compute();
     let handler = config.upload ? config.upload : send;
-    let payload: IEncodedPayload = {e: JSON.stringify(envelope(last)), m: JSON.stringify(metrics(last)), d: `[${events.join()}]`};
-    handler(stringify(payload), last);
-    if (last) { backup(payload); }
+    let payload: EncodedPayload = {e: JSON.stringify(envelope(last)), d: `[${events.join()}]`};
+    let sequence = metadata.envelope.sequence;
+    transit[sequence] = { data: stringify(payload), attempts: 1 };
+    handler(transit[sequence].data, sequence, last);
+    if (last) { backup(payload); } else { ping.reset(); }
     events = [];
-    if (!last) { ping.reset(); }
 }
 
-function stringify(payload: IEncodedPayload): string {
-    return `{"e":${payload.e},"m":${payload.m},"d":${payload.d}}`;
+function stringify(payload: EncodedPayload): string {
+    return `{"e":${payload.e},"d":${payload.d}}`;
 }
 
-function send(data: string, last: boolean = false): void {
+function send(data: string, sequence: number = null, last: boolean = false): void {
     if (config.url.length > 0) {
         if (last && "sendBeacon" in navigator) {
             navigator.sendBeacon(config.url, data);
```

Comment:
>navigator [](start = 12, length = 9)

Future: Should fall back to XHR if this causes an exception -- e.g.., if the payload is too large for a beacon.

Reply:
This is a good point and will think over it before addressing it in the next version.